### PR TITLE
fix issue40

### PR DESCRIPTION
**Pull request by gaoxinge**
**Which issue(s) this PR fixes**:

Fixes https://github.com/dubbogo/hessian2/issues/40

**Does this PR introduce a user-facing change?**:

```release-note
none
```

### CODE REVIEW

**Comment by wongoo** [comment]
@gaoxinge currently there is not a test case for list with nil item, pls add an unit test for that, thanks~

**Comment by wongoo** [comment]
@gaoxinge great work！And pls also add encoding unit test for list with nil item.

**Comment by gaoxinge** [comment]
@wongoo Ok!

**Comment by wongoo** [comment]
@gaoxinge thanks, after adding encode unit test for it, we will merge it asap.

**Comment by gaoxinge** [comment]
- [x] change official hessian to dubbo hessian
- [ ] add encode unit test for list

**Comment by gaoxinge** [comment]
@wongoo I add unit test for encode list and merge master branch.

File: list.go [hunk 1]
```diff
@@ -186,16 +186,15 @@ func (d *Decoder) readTypedList(tag byte) (interface{}, error) {
...
-			break
-		}
-
-		v := EnsureRawValue(it)
 		if isVariableArr {
-			aryValue = reflect.Append(aryValue, v)
+			if it != nil {
+				aryValue = reflect.Append(aryValue, EnsureRawValue(it))
+			} else {
+				aryValue = reflect.Append(aryValue, reflect.Zero(reflect.TypeOf((*interface{})(nil)).Elem()))
```
Comment (wongoo) [inline list.go]: the zero value won't change and could be  a variable

File: list.go [hunk 2]
```diff
@@ -239,7 +238,11 @@ func (d *Decoder) readUntypedList(tag byte) (interface{}, error) {
 		}
 
 		if isVariableArr {
-			aryValue = reflect.Append(aryValue, EnsureRawValue(it))
+			if it != nil {
+				aryValue = reflect.Append(aryValue, EnsureRawValue(it))
+			} else {
+				aryValue = reflect.Append(aryValue, reflect.Zero(reflect.TypeOf((*interface{})(nil)).Elem()))
```
Comment (wongoo) [inline list.go]: as above

File: test_hessian/src/main/java/test/TestHessian.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import com.alibaba.com.caucho.hessian.io.Hessian2Output;
+import com.caucho.hessian.test.A0;
+import com.caucho.hessian.test.A1;
+
+import java.io.OutputStream;
+import java.util.Date;
+import java.util.HashMap;
+
+
+public class TestHessian {
```
Comment (wongoo) [inline test_hessian/src/main/java/test/TestHessian.java]: could be rename to `TestCustomReply`

File: test_hessian/src/main/java/test/TestHessianEncode.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import com.caucho.hessian.io.Hessian2Input;
+import com.caucho.hessian.test.A0;
+import com.caucho.hessian.test.A1;
+
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class TestHessianEncode {
```
Comment (wongoo) [inline test_hessian/src/main/java/test/TestHessianEncode.java]: could be renamed to `TestCustomDecode`

File: encode_test.go [hunk 5]
```diff
@@ -24,3 +25,49 @@ var assertEqual = func(want, got []byte, t *testing.T) {
...
+func encodeTarget(target interface{}) ([]byte, error) {
+	e := NewEncoder()
+	err := e.Encode(target)
+	if err != nil {
+		return nil, err
+	}
+	return e.Buffer(), nil
+}
+
+func getArg(method string, target interface{}) (string, error) {
```
Comment (wongoo) [inline encode_test.go]: could be renamed to `javaDecodeValidate`

File: encode_test.go [hunk 6]
```diff
@@ -24,3 +25,49 @@ var assertEqual = func(want, got []byte, t *testing.T) {
...
+	}
+
+	out, e := cmd.Output()
+	if e != nil {
+		return "", e
+	}
+	return string(out), nil
+}
+
+func testEncodeFramework(t *testing.T, method string, target interface{}) {
```
Comment (wongoo) [inline encode_test.go]: could be renamed to `testJavaDecode`

File: ref.go [hunk 7]
```diff
@@ -25,9 +25,12 @@ import (
 
 // used to ref object,list,map
 type _refElem struct {
-	// record the kind of target, objects are the same only if the address and kind are the same
+	// different kind may share the same address, i.e. slice and its first element
```
Comment (wongoo) [inline ref.go]: the previous comment should not be deleted

File: ref.go [hunk 8]
```diff
@@ -95,13 +99,17 @@ func (e *Encoder) checkRefMap(v reflect.Value) (int, bool) {
 			v = v.Elem()
 		}
 		kind = v.Elem().Kind()
+		if kind != reflect.Invalid {
```
Comment (wongoo) [inline ref.go]: when will kind be Invalid?

File: ref.go [hunk 9]
```diff
@@ -111,15 +119,18 @@ func (e *Encoder) checkRefMap(v reflect.Value) (int, bool) {
 	}
 
 	if elem, ok := e.refMap[addr]; ok {
-		// the array addr is equal to the first elem, which must ignore
 		if elem.kind == kind {
-			return elem.index, ok
+			if elem.kind != reflect.Struct {
```
Comment (wongoo) [inline ref.go]: is it a map or list if not struct?

**Comment by fangyincheng** [comment]
LGTM

File: ref.go [hunk 10]
```diff
@@ -111,15 +119,18 @@ func (e *Encoder) checkRefMap(v reflect.Value) (int, bool) {
...
+				return elem.index, ok
+			} else if elem.tp == tp {
```
Comment (wongoo) [inline ref.go]: the `else if` means the `elem` is  a struct,  when will its type being not equal if it's kind is equal? pls add more comment about this logic, tks

File: ref.go (hunk 8 above)
Comment (gaoxinge) [reply ref.go]: If `v` is `nil`, then `v.Elem().Kind()` equsals `reflect.Invalid` and `v.Elem().Type()` will panic.

File: ref.go (hunk 9 above)
Comment (gaoxinge) [reply ref.go]: Yes, and also may include `reflect.Invalid`.

File: ref.go (hunk 7 above)
Comment (gaoxinge) [reply ref.go]: Ok, I will revert this comment.

File: ref.go (hunk 10 above)
Comment (gaoxinge) [reply ref.go]: - Firstly kind must be equal.
- Secondly there are two situations:
   - if kind is not `reflect.Struct`, just return `elem.index, ok`
   - if kind is `reflet.Struct`, we must ensure the types must equal 

The reason that why we need to use type information to distinguish `struct` is that different empty struct may share the same address and kind.

```go
package main

import (
    "fmt"
    "reflect"
    "unsafe"
)

type A0 struct{}

func (*A0) JavaClassName() string {
    return "com.caucho.hessian.test.A0"
}

type A1 struct{}

func (*A1) JavaClassName() string {
    return "com.caucho.hessian.test.A1"
}

func main() {
    a0 := new(A0)
    a1 := new(A1)
    v1 := reflect.ValueOf(a0)
    v2 := reflect.ValueOf(a1)
    fmt.Println(unsafe.Pointer(a0))  // 0x5761c8
    fmt.Println(unsafe.Pointer(a1))  // 0x5761c8
    fmt.Println(unsafe.Sizeof(a0))   // 8
    fmt.Println(unsafe.Sizeof(a1))   // 8
    fmt.Println(v1.Elem().Kind())    // struct
    fmt.Println(v2.Elem().Kind())    // struct
    fmt.Println(v1.Elem().Type())    // main.A0
    fmt.Println(v2.Elem().Type())    // main.A1
}
```

Futher more, if we don't use the type information to distinguish `struct`, the code below will run wrongly

```go
func TestEmptyStruct(t *testing.T) {
	expected := []interface{}{new(A0), new(A1)}
	
	e := NewEncoder()
	e.Encode(expected)
	
	d := NewDecoder(e.Buffer())
	result, _ := d.Decode()
	
	if !reflect.DeepEqual(expected, result) {
		t.Errorf("encoding wrongly")
	}
}
```

Because `new(A0)` and `new(A1)` share the same address and kind and encoder think they are the same,  then `result` will be `[]interface{}{new(A0), new(A0)}`.

Comment (wongoo) [reply ref.go]: got it, this situation seems only occurs for empty struct, the key description above can be added in code comment.

Comment (gaoxinge) [reply ref.go]: @wongoo I have add comments:

https://github.com/dubbogo/hessian2/blob/ad9cb8199f7324a978de9d46309d13f92ae5b070/ref.go#L31-L33

https://github.com/dubbogo/hessian2/blob/ad9cb8199f7324a978de9d46309d13f92ae5b070/ref.go#L124-L131

File: ref.go (hunk 7 above)
Comment (gaoxinge) [reply ref.go]: I have add the previous comment back:

https://github.com/dubbogo/hessian2/blob/ad9cb8199f7324a978de9d46309d13f92ae5b070/ref.go#L27-L29

**Comment by wongoo** [comment]
@gaoxinge tks for the hard work！
